README.md
# Flask-FomanticUI

[![Build status](https://github.com/juniors90/Flask-FomanticUI/actions/workflows/testing-package.yml/badge.svg)](https://github.com/juniors90/Flask-FomanticUI/actions)
[![codecov](https://codecov.io/gh/juniors90/Flask-FomanticUI/branch/main/graph/badge.svg?token=YNCV9C9GIG)](https://codecov.io/gh/juniors90/Flask-FomanticUI)
![docstr-cov](https://img.shields.io/endpoint?url=https://jsonbin.org/juniors90/Flask-FomanticUI/badges/docstr-cov)
![PyPI](https://img.shields.io/pypi/v/Flask-FomanticUI?logoColor=green)
![PyPI - Downloads](https://img.shields.io/pypi/dm/Flask-FomanticUI)
[![Documentation Status](https://readthedocs.org/projects/flask-fomanticui/badge/?version=latest)](https://flask-fomanticui.readthedocs.io/en/latest/?badge=latest)
[![Forks](https://img.shields.io/github/forks/juniors90/Flask-FomanticUI)](https://github.com/juniors90/Flask-FomanticUI/stargazers)
[![star](https://img.shields.io/github/stars/juniors90/Flask-FomanticUI?color=yellow)](https://github.com/juniors90/Flask-FomanticUI/network/members)
[![issues](https://img.shields.io/github/issues/juniors90/Flask-FomanticUI?color=teal)](https://github.com/juniors90/Flask-FomanticUI/issues)
[![GitHub contributors](https://img.shields.io/github/contributors/juniors90/Flask-FomanticUI?color=green)](https://github.com/juniors90/Flask-FomanticUI/graphs/contributors)
[![black](https://img.shields.io/badge/code%20style-black-000000.svg)](https://github.com/psf/black)

Flask-FomanticUI is a collection of Jinja macros for [Fomantic UI](https://fomantic-ui.com/) and
[Flask](https://flask.palletsprojects.com/en/2.0.x/) for the global style. Very similar
to [Bootstrap-Flask](https://github.com/greyli/bootstrap-flask). 

## Features

- [x] Table generation: Render data objects (dict or class objects) to Fomantic UI Table.
- [x] Paginate generation: Render Flask-SQLAlchemy `Pagination` object to Fomantic UI Pagination.
- [x] Form generation: Render Flask-WTF/WTForms form object to Fomantic UI Form, etc.


## Requirements

Python 3.8+

## Dependecies for this project.

- [Flask(>=2.0.2)](https://flask.palletsprojects.com/en/2.0.x/) for build the backend.

## intallation

You can install via pip:

```cmd
    $> pip install Flask-FomanticUI
```

## Example

Register the extension:

```python
from flask import Flask
# To follow the naming rule of Flask extension, although
# this project's name is Flask-FomanticUI, the actual package
# installed is named `flask_fomanticui`.
from flask_fomanticui import FomanticUI

app = Flask(__name__)
fomantic = FomanticUI(app)
```

Assuming you have a Flask-WTF form like this:

```python
class LoginForm(FlaskForm):
    username = StringField('Username', validators=[DataRequired(), Length(1, 20)])
    password = PasswordField('Password', validators=[DataRequired(), Length(8, 150)])
    submit = SubmitField()
    remember = BooleanField('Remember me')
```

Now with the `render_ui_form` macro:

```html
{% from 'fomanticui/form_ui.html' import render_ui_form %}
<html>
<head>
<!-- Fomantic UI - CSS -->
</head>
<body>

<h2>Login</h2>
{{ render_ui_form(form) }}

<!-- Fomantic UI - JS -->
</body>
</html>
```

You will get a form like this with only one line code (i.e. `{{ render_ui_form(form) }}`):

![form rendering](https://raw.githubusercontent.com/juniors90/Flask-FomanticUI/master/docs/source/_static/form-example.PNG)

When the validation fails, the error messages will be rendered with proper style:

![error form rendering](https://raw.githubusercontent.com/juniors90/Flask-FomanticUI/master/docs/source/_static/error-form-example.PNG)

Read the [Basic Usage](https://flask-fomanticui.readthedocs.io/en/latest/notes/basic.html) 
docs for more details.

## Links

- [Documentation](https://flask-fomanticui.readthedocs.io)
- [Example Application](https://github.com/juniors90/Flask-FomanticUI/tree/main/sample_app)
- [PyPI Releases](https://pypi.org/project/Flask-FomanticUI/)
- [Changelog](https://github.com/juniors90/Flask-FomanticUI/blob/main/CHANGELOG.rst)


## Authors

- Ferreira, Juan David

Please submit bug reports, suggestions for improvements and patches via
the (E-mail: juandavid9a0@gmail.com).

## Contributors

Credits goes to these peoples:

<a href="https://github.com/juniors90/Flask-FomanticUI/graphs/contributors">
  <img src="https://contrib.rocks/image?repo=juniors90/Flask-FomanticUI" />
</a>

## Official repository and Issues

- https://github.com/juniors90/Flask-FomanticUI


## License

`Flask-FomanticUI` is free software you can redistribute it and/or modify it
under the terms of the MIT License. For more information, you can see the
[LICENSE](https://github.com/juniors90/Flask-FomanticUI/blob/main/LICENSE) file
for details.


CHANGELOG.rst
Changelog
=========


0.0.3
-----

- Bumped to FomanticUI 2.9.0 #2


tests/conftest.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

# This file was part of Flask-Bootstrap and was modified under the terms of
# its BSD License. Copyright (c) 2013, Marc Brinkmann. All rights reserved.
#
# This file was part of Bootstrap-Flask and was modified under the terms of
# its MIT License. Copyright (c) 2018 Grey Li. All rights reserved.
#
# This file is part of the
# Flask-FomanticUI Project (https://github.com/juniors90/Flask-FomanticUI/).
# Copyright (c) 2021, Ferreira Juan David
# License: MIT
# Full Text: https://github.com/juniors90/Flask-FomanticUI/blob/master/LICENSE

# =============================================================================
# DOCS
# =============================================================================

import typing as t

import flask
from flask import render_template_string

from flask_wtf import FlaskForm

import pytest as pt

from wtforms import (
    BooleanField,
    HiddenField,
    IntegerField,
    PasswordField,
    RadioField,
    StringField,
    SubmitField,
)
from wtforms.validators import DataRequired, Length


class ExampleForm(FlaskForm):
    name = StringField("Name")
    username = StringField(
        "Username", validators=[DataRequired(), Length(1, 20)]
    )
    password = PasswordField(
        "Password", validators=[DataRequired(), Length(8, 150)]
    )
    country_code = IntegerField("Country Code", validators=[DataRequired()])
    radio_field = RadioField(
        "This is a radio field",
        choices=[
            ("head_radio", "Head radio"),
            ("radio_76fm", "Radio '76 FM"),
            ("lips_106", "Lips 106"),
            ("wctr", "WCTR"),
        ],
    )
    hidden = HiddenField()
    remember = BooleanField("Remember me")
    submit_button = SubmitField("Submit Form")


class HelloForm(FlaskForm):
    name = StringField("Name")
    username = StringField(
        "Username", validators=[DataRequired(), Length(1, 20)]
    )
    password = PasswordField(
        "Password", validators=[DataRequired(), Length(8, 150)]
    )
    remember = BooleanField("Remember me")
    hidden = HiddenField()
    submit = SubmitField()


if t.TYPE_CHECKING:
    from flask.testing import FlaskClient


@pt.fixture
def hello_form():
    return HelloForm


@pt.fixture
def example_form():
    return ExampleForm


@pt.fixture(autouse=True)
def app() -> "flask.Flask":
    app = flask.Flask(__name__)
    app.testing = True
    app.config["SQLALCHEMY_DATABASE_URI"] = "sqlite:///"
    app.config["SQLALCHEMY_TRACK_MODIFICATIONS"] = False
    app.secret_key = "for test"
    app.debug = False

    @app.route("/")
    def index():
        return render_template_string(
            "{{ semantic.load_css() }}{{ semantic.load_js() }}"
        )

    yield app


@pt.fixture
def client(app: "flask.Flask") -> "FlaskClient":
    context = app.test_request_context()
    context.push()

    with app.test_client() as client:
        yield client

    context.pop()


tests/tests_fomantic/conftest.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

# This file was part of Flask-Bootstrap and was modified under the terms of
# its BSD License. Copyright (c) 2013, Marc Brinkmann. All rights reserved.
#
# This file was part of Bootstrap-Flask and was modified under the terms of
# its MIT License. Copyright (c) 2018 Grey Li. All rights reserved.
#
# This file is part of the
# Flask-FomanticUI Project (https://github.com/juniors90/Flask-FomanticUI/).
# Copyright (c) 2021, Ferreira Juan David
# License: MIT
# Full Text: https://github.com/juniors90/Flask-FomanticUI/blob/master/LICENSE

# =============================================================================
# TESTS
# =============================================================================

import re

from flask_fomanticui import FomanticUI

import pytest as pt


@pt.fixture(autouse=True)
def fomantic(app):
    yield FomanticUI(app)


@pt.fixture
def cdn_suiv():
    fomantic_version = re.search(
        r"(\d+\.\d+\.\d+)", str(FomanticUI.fomantic_version)
    ).group(1)
    return "Fomantic UI - " + fomantic_version


@pt.fixture
def cdn_jqv():
    jquery_version = re.search(
        r"(\d+\.\d+\.\d+)", str(FomanticUI.jquery_version)
    ).group(1)
    return "jQuery v" + jquery_version


@pt.fixture
def local_suiv():
    fomantic_version = re.search(
        r"(\d+\.\d+\.\d+)", str(FomanticUI.fomantic_version)
    ).group(1)
    return "Fomantic UI - " + fomantic_version


@pt.fixture
def local_jsv():
    js_version = re.search(
        r"(\d+\.\d+\.\d+)", str(FomanticUI.fomantic_version)
    ).group(1)
    return "Fomantic UI - " + js_version


@pt.fixture
def local_jqv():
    jq_version = re.search(
        r"(\d+\.\d+\.\d+)", str(FomanticUI.jquery_version)
    ).group(1)
    return "jQuery v" + jq_version


tests/tests_fomantic/test_sample.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

# This file was part of Flask-Bootstrap and was modified under the terms of
# its BSD License. Copyright (c) 2013, Marc Brinkmann. All rights reserved.
#
# This file was part of Bootstrap-Flask and was modified under the terms of
# its MIT License. Copyright (c) 2018 Grey Li. All rights reserved.
#
# This file is part of the
# Flask-FomanticUI Project (https://github.com/juniors90/Flask-FomanticUI/).
# Copyright (c) 2021, Ferreira Juan David
# License: MIT
# Full Text: https://github.com/juniors90/Flask-FomanticUI/blob/master/LICENSE

# =====================================================================
# TESTS
# =====================================================================


# =============================================================================
# TEST SAMPLE
# =============================================================================


def test_sample_request(app, client):
    @app.get("/sample")
    def sample():
        return "OK"

    r = client.get("/sample")
    assert r.status_code == 200
    assert r.data == b"OK"


flask_fomanticui/core.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

# This file was part of Flask-Bootstrap and was modified under the terms of
# its BSD License. Copyright (c) 2013, Marc Brinkmann. All rights reserved.
#
# This file was part of Bootstrap-Flask and was modified under the terms of
# its MIT License. Copyright (c) 2018 Grey Li. All rights reserved.
#
# This file is part of the
# Flask-FomanticUI Project (https://github.com/juniors90/Flask-FomanticUI/).
# Copyright (c) 2021, Ferreira Juan David
# License: MIT
# Full Text: https://github.com/juniors90/Flask-FomanticUI/blob/master/LICENSE

# =============================================================================
# DOCS
# =============================================================================

"""Flask-FomanticUI.

Implementation of Fomantic UI in Flask.
"""

# =============================================================================
# IMPORTS
# =============================================================================


import warnings

from flask import Blueprint, Markup, current_app, url_for

try:  # pragma: no cover
    from wtforms.fields import HiddenField
except ImportError:  # pragma: no cover
    # docstr-coverage:excused `no one is reading this anyways`
    def is_hidden_field_filter(field):
        raise RuntimeError("WTForms is not installed.")

else:
    # docstr-coverage:excused `no one is reading this anyways`
    def is_hidden_field_filter(field):
        return isinstance(field, HiddenField)


cdn_base = "https://cdn.jsdelivr.net/npm"


# docstr-coverage:excused `no one is reading this anyways`
def raise_helper(message):  # pragma: no cover
    raise RuntimeError(message)


def get_table_titles(data, primary_key, primary_key_title):
    """Detect and build the table titles tuple from ORM object.

    .. note::
        Currently only support SQLAlchemy.
    """
    if not data:
        return []
    titles = []
    for k in data[0].__table__.columns.keys():
        if not k.startswith("_"):
            titles.append((k, k.replace("_", " ").title()))
    titles[0] = (primary_key, primary_key_title)
    return titles


# docstr-coverage:excused `no one is reading this anyways`
def link_css_with_sri(url, sri):
    return f'<link rel="stylesheet" href="{url}" integrity="{sri}" crossorigin="anonymous">'  # noqa: E501


# docstr-coverage:excused `no one is reading this anyways`
def simple_link_css(url):
    return f'<link rel="stylesheet" href="{url}">'


# docstr-coverage:excused `no one is reading this anyways`
def scripts_with_sri(url, sri):
    return f'<script src="{url}" integrity="{sri}" crossorigin="anonymous"></script>'  # noqa: E501


# docstr-coverage:excused `no one is reading this anyways`
def simple_scripts_js(url):
    return f'<script src="{url}"></script>'


class _FomanticUI(object):
    """Base extension class for different Fomantic UI versions.

    .. versionadded:: 0.0.1
    """

    fomantic_version = None
    jquery_version = None
    fomantic_css_integrity = None
    fomantic_js_integrity = None
    jquery_integrity = None
    static_folder = None
    fomantic_css_filename = "semantic.min.css"
    fomantic_js_filename = "semantic.min.js"
    jquery_filename = "jquery.min.js"

    def __init__(self, app=None):
        if app is not None:
            self.init_app(app)

    def init_app(self, app):
        """Application factory."""

        # default settings
        app.config.setdefault("FOMANTIC_SERVE_LOCAL", False)
        app.config.setdefault("FOMANTIC_BUTTON_STYLE", "primary")
        app.config.setdefault("FOMANTIC_BUTTON_SIZE", "")
        app.config.setdefault("FOMANTIC_ICON_COLOR", None)
        app.config.setdefault("FOMANTIC_ICON_SIZE", None)
        app.config.setdefault(
            "FOMANTIC_ERROR_HEADER", "Action Forbidden"
        )  # noqa: E501
        app.config.setdefault(
            "FOMANTIC_CHECKBOX_HEADER_ERROR", "Action Forbidden"
        )  # noqa: E501
        app.config.setdefault(
            "FOMANTIC_RADIO_HEADER_ERROR", "Action Forbidden"
        )  # noqa: E501
        app.config.setdefault(
            "FOMANTIC_MSG_CATEGORY", None
        )  # change primary by None
        app.config.setdefault("FOMANTIC_TABLE_VIEW_TITLE", "View")
        app.config.setdefault("FOMANTIC_TABLE_EDIT_TITLE", "Edit")
        app.config.setdefault("FOMANTIC_TABLE_DELETE_TITLE", "Delete")
        app.config.setdefault("FOMANTIC_TABLE_NEW_TITLE", "New")
        if not hasattr(app, "extensions"):  # pragma: no cover
            app.extensions = {}
        app.extensions["fomantic"] = self
        blueprint = Blueprint(
            "fomantic",
            __name__,
            static_folder=f"static/{self.static_folder}",
            static_url_path=f"{app.static_url_path}",
            template_folder="templates",
        )

        app.register_blueprint(blueprint)

        app.jinja_env.globals["fomantic"] = self
        app.jinja_env.globals[
            "fomantic_is_hidden_field"
        ] = is_hidden_field_filter
        app.jinja_env.globals["get_table_titles"] = get_table_titles
        app.jinja_env.globals["warn"] = warnings.warn
        app.jinja_env.globals["raise"] = raise_helper
        app.jinja_env.add_extension("jinja2.ext.do")

    def load_css(self, s_version=None, fomantic_sri=None):
        """Load Fomantic's css resources with given version.

        Parameters
        ----------
        s_version : str
            The version of Fomantic UI.
        fomantic_sri : str
            Subresource Integrity.
        Return
        ------
        scripts_cdn : markupsafe.Markup
            Fomantic UI CDN File.
        """

        serve_local = current_app.config["FOMANTIC_SERVE_LOCAL"]
        s_version = self.fomantic_version if s_version is None else s_version
        fomantic_sri = self._get_sri("fomantic_css", s_version, fomantic_sri)

        if serve_local:
            base_path = "css"
            url = url_for(
                "fomantic.static",
                filename=f"{base_path}/{self.fomantic_css_filename}",
            )
        else:
            base_path = cdn_base + f"/fomantic-ui@{s_version}/dist/"
            url = base_path + self.fomantic_css_filename

        if fomantic_sri:
            css = link_css_with_sri(url, fomantic_sri)
        else:
            css = f'<link rel="stylesheet" type="text/css" href="{url}">'
        scripts_cdn = Markup(css)
        return scripts_cdn

    def _get_js_script(self, version, name, sri):
        """Get <script> tag for JavaScipt resources."""
        serve_local = current_app.config["FOMANTIC_SERVE_LOCAL"]
        paths = {
            "fomantic-ui": f"{self.fomantic_js_filename}",
            "jquery": f"{self.jquery_filename}",
        }

        if serve_local:
            path = "js/semantic"
            url = url_for("fomantic.static", filename=f"{path}/{paths[name]}")
        else:
            url = cdn_base + f"/{name}@{version}/dist/{paths[name]}"

        if sri:
            script_html = scripts_with_sri(url, sri)
        else:
            script_html = simple_scripts_js(url)
        return script_html

    def _get_sri(self, name, version, sri):
        serve_local = current_app.config["FOMANTIC_SERVE_LOCAL"]

        sris = {
            "fomantic_css": self.fomantic_css_integrity,
            "fomantic_js": self.fomantic_js_integrity,
            "jquery": self.jquery_integrity,
        }

        versions = {
            "fomantic_css": self.fomantic_version,
            "fomantic_js": self.fomantic_version,
            "jquery": self.jquery_version,
        }

        if sri is not None:
            return sri
        if version == versions[name] and serve_local is False:
            return sris[name]
        return None

    def load_js(
        self,
        version=None,
        jq_version=None,  # noqa: C901
        fomantic_sri=None,
        jquery_sri=None,
    ):
        """Load Fomantic UI and other resources with given version.

        Parameter
        ---------
        version : str
            The version of Fomantic UI.
        jq_version : str
            The version of jQuery (for Fomantic UI).
        fomantic_sri : str
            Subresource Integrity for Fomantic UI..
        jquery_sri : str
            Subresource Integrity for jQuery.
        Return
        ------
        scripts_cdn : markupsafe.Markup
            Fomantic UI CDN File.
        """

        fui_version = self.fomantic_version if version is None else version
        fui_sri = self._get_sri("fomantic_js", fui_version, fomantic_sri)
        fui_js = self._get_js_script(fui_version, "fomantic-ui", fui_sri)
        jq_version = self.jquery_version if jq_version is None else jq_version
        jquery_sri = self._get_sri("jquery", jq_version, jquery_sri)
        jquery = self._get_js_script(jq_version, "jquery", jquery_sri)
        scripts_cdn = Markup(
            f"""{jquery}
                {fui_js}"""
        )
        return scripts_cdn


flask_fomantic/fomantic.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

# This file was part of Flask-Bootstrap and was modified under the terms of
# its BSD License. Copyright (c) 2013, Marc Brinkmann. All rights reserved.
#
# This file was part of Bootstrap-Flask and was modified under the terms of
# its MIT License. Copyright (c) 2018 Grey Li. All rights reserved.
#
# This file is part of the
# Flask-FomanticUI Project (https://github.com/juniors90/Flask-FomanticUI/).
# Copyright (c) 2021, Ferreira Juan David
# License: MIT
# Full Text: https://github.com/juniors90/Flask-FomanticUI/blob/master/LICENSE

# =============================================================================
# DOCS
# =============================================================================

"""Flask-FomanticUI.

Implementation of FomanticUI in Flask.
"""

# =============================================================================
# META
# =============================================================================

from .core import _FomanticUI


class FomanticUI(_FomanticUI):
    """Base class for palies the FomanticUI framework.

    Initilize the extension::

        from flask import Flask
        from flask_fomantic import FomanticUI

        app = Flask(__name__)
        fomantic = FomanticUI(app)


    Or with the application factory::


        from flask import Flask
        from flask_fomantic import FomanticUI

        Fomantic = FomanticUI()

        def create_app():
            app = Flask(__name__)
            fomantic.init_app(app)

    """

    jquery_version = "3.3.1"
    fomantic_version = "2.8.8"
    fomantic_css_integrity = (
        "sha256-ckBg2pN9ZF9KPn+vY00JXAkdqE1nv20iGM2haZhCBl4="
    )
    fomantic_js_integrity = (
        "sha256-VxL9ZXOItJ0i4nJLm39HIoX8u3cCRPRkDjMSXZ/RiQQ="
    )
    jquery_integrity = "sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8="
    static_folder = "fomantic"


flask_fomanticui/__init__.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

# This file was part of Flask-Bootstrap and was modified under the terms of
# its BSD License. Copyright (c) 2013, Marc Brinkmann. All rights reserved.
#
# This file was part of Bootstrap-Flask and was modified under the terms of
# its MIT License. Copyright (c) 2018 Grey Li. All rights reserved.
#
# This file is part of the
# Flask-FomanticUI Project (https://github.com/juniors90/Flask-FomanticUI/).
# Copyright (c) 2021, Ferreira Juan David
# License: MIT
# Full Text: https://github.com/juniors90/Flask-FomanticUI/blob/master/LICENSE

# =============================================================================
# DOCS
# =============================================================================

"""Flask-FomanticUI.

Implementation of FomanticUI in Flask.
"""

# =============================================================================
# META
# =============================================================================

__version__ = "0.0.3"

# =============================================================================
# IMPORTS
# =============================================================================

from .core import *  # noqa

from .fomantic import *  # noqa


sample_app/fomantic/app.py
# -*- coding: utf-8 -*-
import os
import pathlib
import sys

from flask import (
    Flask,
    Markup,
    flash,
    redirect,
    render_template,
    render_template_string,
    request,
    url_for,
)

from flask_sqlalchemy import SQLAlchemy

from flask_wtf import CSRFProtect, FlaskForm

from wtforms import (
    BooleanField,
    FormField,
    IntegerField,
    PasswordField,
    StringField,
    SubmitField,
)
from wtforms.fields import FileField, HiddenField, RadioField
from wtforms.validators import DataRequired, Length


# this path is pointing to project/docs/source
CURRENT_PATH = pathlib.Path(os.path.abspath(os.path.dirname(__file__)))
FLASK_FOMANTIC_PATH = CURRENT_PATH.parent.parent

sys.path.insert(0, str(FLASK_FOMANTIC_PATH))

from flask_fomanticui import FomanticUI  # noqa

app = Flask(__name__)
# app.secret_key = 'dev'
app.config["SECRET_KEY"] = "secret-key"
app.config["FOMANTIC_SERVE_LOCAL"] = True

app.config["SQLALCHEMY_TRACK_MODIFICATIONS"] = False
app.config["SQLALCHEMY_DATABASE_URI"] = "sqlite:///:memory:"

# set default button sytle and size, will be overwritten by macro parameters
app.config["FOMANTIC_BUTTON_STYLE"] = "primary"
app.config["FOMANTIC_BUTTON_SIZE"] = ""

# set default icon title of table actions
app.config["FOMANTIC_TABLE_VIEW_TITLE"] = "Read"
app.config["FOMANTIC_TABLE_EDIT_TITLE"] = "Update"
app.config["FOMANTIC_TABLE_DELETE_TITLE"] = "Remove"
app.config["FOMANTIC_TABLE_NEW_TITLE"] = "Create"

app.config["FOMANTIC_CHECKBOX_HEADER_ERROR"] = "Checkbox Header Error"
app.config["FOMANTIC_RADIO_HEADER_ERROR"] = "Radio Header Error"

semantic = FomanticUI(app)
db = SQLAlchemy(app)
csrf = CSRFProtect(app)


# straight from the wtforms docs:
class TelephoneForm(FlaskForm):
    country_code = IntegerField("Country Code", validators=[DataRequired()])
    area_code = IntegerField("Area Code/Exchange", validators=[DataRequired()])
    number = StringField(
        "Number",
        validators=[
            DataRequired(message="My custom message for number field."),
            Length(min=6, max=14),
        ],
    )


class ExampleForm(FlaskForm):
    field1 = StringField(
        "First Field",
        description="This is field one.",
        validators=[DataRequired()],
    )
    field2 = StringField(
        "Second Field",
        description="This is field two.",
        validators=[DataRequired()],
    )
    hidden_field = HiddenField("You cannot see this", description="Nope")
    radio_field = RadioField(
        "This is a radio field",
        choices=[
            ("head_radio", "Head radio"),
            ("radio_76fm", "Radio '76 FM"),
            ("lips_106", "Lips 106"),
            ("wctr", "WCTR"),
        ],
        validators=[DataRequired()],
    )
    checkbox_field = BooleanField(
        "This is a checkbox",
        description="Checkboxes can be tricky.",
        validators=[DataRequired()],
    )

    # subforms
    mobile_phone = FormField(TelephoneForm)

    # you can change the label as well
    office_phone = FormField(TelephoneForm, label="Your office phone")

    ff = FileField("Sample upload", validators=[DataRequired()])

    submit_button = SubmitField("Submit Form")


class Message(db.Model):
    id = db.Column(db.Integer, primary_key=True)  # noqa: A003
    text = db.Column(db.Text, nullable=False)
    author = db.Column(db.String(100), nullable=False)
    category = db.Column(db.String(100), nullable=False)
    draft = db.Column(db.Boolean, default=False, nullable=False)
    create_time = db.Column(db.Integer, nullable=False, unique=True)


@app.before_first_request
def before_first_request_func():
    db.drop_all()
    db.create_all()
    for i in range(20):
        m = Message(
            text=f"Test message {i+1}",
            author=f"Author {i+1}",
            category=f"Category {i+1}",
            create_time=4321 * (i + 1),
        )
        if i % 4:
            m.draft = True
        db.session.add(m)
    db.session.commit()


@app.route("/")
def index():
    return render_template("index.html")


@app.route("/form", methods=["GET", "POST"])
def test_form():
    form = ExampleForm()
    return render_template("form_basic.html", form=form)


@app.route("/form-inline/", methods=["GET", "POST"])
def test_form_inline():
    form = ExampleForm()
    return render_template("form_inline.html", form=form)


@app.route("/form-inverted", methods=["GET", "POST"])
def test_form_inverted():
    form = ExampleForm()
    if form.validate_on_submit():
        name = form.field1.data  # noqa: F841
        # .. other fields
        next = request.args.get("next", None)  # noqa: A001
        if next:
            return redirect(next)
        return redirect(url_for("index"))
    return render_template("form_inverted.html", form=form)


@app.route("/form-inline-inverted", methods=["GET", "POST"])
def test_form_inline_inverted():
    form = ExampleForm()
    return render_template("form_inline_inverted.html", form=form)


@app.route("/nav", methods=["GET", "POST"])
def test_nav():
    form = ExampleForm()
    return render_template("nav.html", form=form)


@app.route("/pagination", methods=["GET", "POST"])
def test_pagination():
    page = request.args.get("page", 1, type=int)
    pagination = Message.query.paginate(page, per_page=2)
    messages = pagination.items
    return render_template(
        "pagination.html", pagination=pagination, messages=messages
    )


@app.route("/flash", methods=["GET", "POST"])
def test_flash():
    flash("A simple default alert—check it out!")
    for category in [
        "dark",
        "danger",
        "debug",
        "light",
        "critical",
        "error",
        "info",
        "warning",
        "success",
    ]:
        flash(f"A simple {category} alert—check it out!", category)
    return render_template("flash.html")


@app.route("/tables/")
def test_table():
    page = request.args.get("page", 1, type=int)
    pagination = Message.query.paginate(page, per_page=10)
    messages = pagination.items
    titles = [
        ("id", "#"),
        ("text", "Message"),
        ("author", "Author"),
        ("category", "Category"),
        ("draft", "Draft"),
        ("create_time", "Create Time"),
    ]
    return render_template(
        "table.html", messages=messages, titles=titles, Message=Message
    )


@app.route("/tables/<int:message_id>/view")
def view_message(message_id):
    message = Message.query.get(message_id)
    not_exist_msg = (
        f"Could not view message {message_id}"
        + " as it does not exist. Return to "
        + '<a href="/tables">table</a>.'
    )
    exist_msg = (
        f"Viewing {message_id}"
        + ' with text "{message.text}".'
        + ' Return to <a href="/tables">table</a>.'
    )
    if message:
        return exist_msg
    return not_exist_msg


@app.route("/tables/<int:message_id>/edit")
def edit_message(message_id):
    message = Message.query.get(message_id)
    not_exist_msg = (
        f"Message {message_id} did not exist and could therefore"
        + ' not be edited. Return to <a href="/tables">table</a>.'
    )
    if message:
        message.draft = not message.draft
        db.session.commit()
        return (
            f"Message {message_id} has been editted by toggling draft status."
            + ' Return to <a href="/tables">table</a>.'
        )
    return not_exist_msg


@app.route("/tables/<int:message_id>/delete", methods=["POST"])
def delete_message(message_id):
    message = Message.query.get(message_id)
    if message:
        db.session.delete(message)
        db.session.commit()
        return (
            f"Message {message_id} has been deleted. Return to"
            + ' <a href="/tables">table</a>.'
        )
    return (
        f"Message {message_id} did not exist and could therefore not be"
        + ' deleted. Return to <a href="/tables">table</a>.'
    )


@app.route("/tables/<int:message_id>/like")
def like_message(message_id):
    return (
        f"Liked the message {message_id}. Return to "
        + '<a href="/tables">table</a>.'
    )


@app.route("/tables/new-message")
def new_message():
    return (
        'Here is the new message page. Return to <a href="/tables">table</a>.'
    )


class Msg(db.Model):
    id = db.Column(db.Integer, primary_key=True)  # noqa: A003
    text = db.Column(db.Text)


@app.route("/test-render-field")
def test_field():
    form = ExampleForm()
    return render_template_string(
        """
        {% extends 'base.html' %}
        {% from 'fomantic/form_ui.html' import render_ui_field,
                                               render_ui_hidden_errors %}
        {% block content %}
        <h1 class="ui center">Flask Semantic UI 2.4.2 - Demo Application</h1>
        <h2>This is the sample inline form</h2>
            <form class="ui form error" method="post">
                {{ form.hidden_tag() }}
                {{ render_ui_hidden_errors(form) }}
                {{ render_ui_field(form.username) }}
                {{ render_ui_field(form.password) }}
                {{ render_ui_field(form.submit) }}
            </form>
        {% endblock %}""",
        form=form,
    )


@app.route("/icon")
def test_icon():
    return render_template("icon.html")


@app.route("/testflash")
def test_test():
    flash(
        Markup('a info message with a link: <a href="/">Click me!</a>'), "info"
    )  # noqa: E501
    return render_template("flash.html")


class LoginForm(FlaskForm):
    username = StringField(
        "Username", validators=[DataRequired(), Length(1, 20)]
    )
    password = PasswordField(
        "Password", validators=[DataRequired(), Length(8, 150)]
    )
    submit = SubmitField()
    remember = BooleanField("Remember me")


@app.route("/login", methods=["GET", "POST"])
def test_login():
    form = LoginForm()
    if form.validate_on_submit():
        name = form.username.data  # noqa: F841
        password = form.password.data  # noqa: F841
        remember = form.remember.data  # noqa: F841
        next = request.args.get("next", None)  # noqa: A001
        if next:
            return redirect(next)
        return redirect(url_for("index"))
    return render_template("login.html", form=form)


if __name__ == "__main__":
    app.run(debug=True)
